macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+
+import macaroons
+import requests
+
+
+def deserialize(json_macaroon):
+    '''Deserialize a JSON macaroon into a macaroon object.
+
+    @param the JSON macaroon to deserialize as a dict.
+    @return the deserialized macaroon object.
+    '''
+    return macaroons.deserialize(to_base64(json_macaroon))
+
+
+def serialize_macaroon_string(macaroon):
+    '''Serialize macaroon object to string.
+
+    @param macaroon object to be serialized.
+    @return a string serialization form of the macaroon.
+    '''
+    return base64.urlsafe_b64decode(
+        _add_padding(macaroon.serialize_json()))
+
+
+def to_base64(json_macaroon):
+    '''Base 64 encode a JSON macaroon.
+
+    @param the JSON macaroon as a dict to encode.
+    @return the encoded macaroon.
+    '''
+    return base64.urlsafe_b64encode(json.dumps(json_macaroon))
+
+
+def _add_padding(s):
+    '''Add padding to base64 encoded string
+    libmacaroons does not give padded base64 string from serialization.
+
+    @param string s to be padded.
+    @return a padded string.
+    '''
+    return s + '=' * ((4 - (len(s) % 4)) % 4)
+
+
+class DischargeError(Exception):
+    '''Exception that is been raised by Bakery public methods when a discharge
+       error happens.
+    '''
+    pass
+
+
+class Bakery(object):
+    ''' Bakery is a wrapper around requests that ensures
+        that macaroons headers are sent to the server if present or gather the
+        discharged macaroons automatically before sending the request.
+    '''
+    def __init__(self, initial_macaroons=None, discharge_token=None,
+                 timeout=30):
+        '''
+        @param initial_macaroons initial macaroon to be used in headers as a
+            base 64 encoded array of base64 encoded json(dict) macaroons.
+        @param discharge_token initial discharge token
+            as a base64 encoded json (dict) macaroon.
+        @param timeout timeout for wait url.
+        '''
+        self._macaroons = initial_macaroons
+        self._discharge_token = discharge_token
+        self._timeout = timeout
+
+    def get(self, url, **kwargs):
+        '''Act as the same than requests.get, same arguments,
+           same returned value, same exception raised except DischargeError
+           when something wrong happened when discharging.
+           If macaroons are present, the headers will be augmented with it.
+        '''
+        if self._macaroons is not None:
+            headers = kwargs.setdefault('headers', {})
+            headers.update({
+                'Bakery-Protocol-Version': 1,
+                'Macaroons': self._macaroons,
+            })
+        response = requests.get(url, **kwargs)
+        if (response.status_code == 401 and
+                response.headers['Www-Authenticate'] == 'Macaroon') \
+                or response.status_code == 407:
+            try:
+                response_json = response.json()
+            except ValueError:
+                raise DischargeError(
```

Comment:
Wouldn't it be better to include the original ValueError in the exception? (or perhaps it is included anyway?)

Also, I'm not sure whether DischargeError is appropriate here. In the Go bakery, at any rate, DischargeError is returned when a third party refuses a discharge, which isn't the case here - we've just got a badly behaved server.